### ISSUE TITLE
Master job (suggested feature)

### ISSUE THREAD

**Issue by ManuelCalavera** [zeagler/tangerine#34]
@zeagler I'm still thinking about this one, but wanted to get an idea out there.

The notion goes back to improving the experience for breaking out a large job into multiple small jobs.

The thought is basically:
> If I customer job truly consists of a multiple of smaller (sub) jobs, I don't necessarily want to see these sub jobs as scheduled unless the job:
> - is awaiting execution
> - is currently executing
> - has errored
> - is being edited
> 
> The idea further extends to making maintenance and managing (including editing) these type of jobs easier.

So, conceptually I decided to draw up a job schedule process in YAML (see below):
```
masterjob:
  -
    name: Cust1-MasterJob
    prefix: Cust1
    cron: "0 5 * * *"
    defaultEnvironment:
      - "JAVA_OPTIONS = -Xmx26g"
      - "APP_FILES = /app-files/config-files"
    defaultDataVolumes:
      - "/home/ubuntu/shared/folder1:/folder1"
      - "/home/ubuntu/shared/folder1/config-files/folder2:/home/app/folder2"
    defaultImage:
      - registry/user/image:1.1
  -
    jobs:
      -
        jobname: PreProcess
        jobtype: container
        command: kitchen.sh -file=/PreProcessJob.kjb -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_PreProcess_$$date_$$time.log
      -
        jobname: Extract1
        jobtype: container
        jobtag: extract
        command: kitchen.sh -file=/PreProcessJob.kjb -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Extract1_$$date_$$time.log
        dependencies:
          - PreProcess
      -
        jobname: Extract2
        jobtype: container
        jobtag: extract
        command: kitchen.sh -file=/PreProcessJob.kjb -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Extract2_$$date_$$time.log
        dependencies:
          - PreProcess
      -
        jobname: Dimension1
        jobtype: container
        jobtag: dimension
        command: kitchen.sh -file=/Dimension_Job.kjb -param:artifactname=Dimension1 -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Dimension1_$$date_$$time.log
        dependencies:
          - "tag:extract"
      -
        jobname: Dimension2
        jobtype: container
        jobtag: dimension
        command: kitchen.sh -file=/Dimension_Job.kjb -param:artifactname=Dimension2 -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Dimension2_$$date_$$time.log
        dependencies:
          - "tag:extract"
      -
        jobname: Dimension3
        jobtype: container
        jobtag: dimension
        command: kitchen.sh -file=/Dimension_Job.kjb -param:artifactname=Dimension3 -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Dimension3_$$date_$$time.log
        dependencies:
          - Dimension1
      -
        jobname: Fact1
        jobtype: container
        jobtag: fact
        command: kitchen.sh -file=/Fact_Job.kjb -param:artifactname=Fact1  -param:propertiesPath=/properties/Cust1.properties -logfile=/logs/Cust1/Cust1_Fact1_$$date_$$time.log
        dependencies:
          - "tag:dimension"
```

Now, the above is a thought to start a conversation. It doesn't include several of the fields you've already added that are still very much useful. Instead, the above YAML sample includes only the bare minimum for this particular conversation.

In the above outline, there are a few thoughts and then some things implied:
- The "sub" jobs essentially rely on the schedule of the "master job". In other words, the sub jobs do not need their own cron.
- The "sub jobs" are actually created for each run. So when the "master job" starts it will create entries for all sub jobs (making them visible on the UI), and when the job completes (successfully) the "sub jobs" will be deleted.
- If an error occurs the outstanding sub jobs will still show as pending execution (dependencies), except for the errored job. Once no more executions can occur (because they execute in parallel), the main job will set itself to errored as well.
- Stopping the main job will also stop the sub jobs and then delete these sub jobs.
- Dimension dependencies in the above work in two different ways:
   - By job (just as you have today)
   - By tag. This prevents having to list out every dimension job for a fact run, but instead we can simply reference `tag:dimension` (making maintenance easier when we add/remove jobs)
- Certain job specifications can be inherited from the master job level (default)

What do you think? Not saying this (YAML in particular) is necessarily the solution, but this is more from making it easier.

**Comment by zeagler** [comment]
This feature is next in line for development, @ManuelCalavera @HarikaVuppala since you two are the primary users we can start solidifying the idea so that it works best for everyones needs. Take note that overtime the user interface will probably take on a very different look and feel that makes some of these issues make more sense visually.

To supplement the above suggestions:
* A `Master Job`, or maybe just `Job`, is a series of `Tasks` that share a configuration and schedule
* The tasks can exist but not be visible unless the user wants to see them
* A UI component to "drill down" into a master job to see and manage the tasks when the job is not being executed
* The master job has a default configuration that can be overridden at the task level
* A failed task should not stop the job, just block any tasks that depend on it.
* A manual start of the job should have an option to "resume" where it failed or "restart" to run every task.
* A scheduled start should "restart" all the tasks unless the user sets an option to wait until the error is reviewed.
* The history tab should combine all tasks into one `Run` and provide metrics and statistics for each individual `Task` as well as the entire `Job`. Information for the individual tasks would be provided as the user requests them.
* Jobs should be nestable. (long term feature)
```
Big Job
    Little Job 1
        Task A1
        Task A2
    Little Job 2 - depends on Little Job 1
        Task B1
        Task B2
    Validation Task - depends on Little Job 2
```
As per #21, I hope that JSON, YAML, and CSV can be used interchangeably by preference for Tangerine.

**Comment by ManuelCalavera** [comment]
@zeagler most of the above looks great. Let's stick with the `Job` and `Task` notation hereon out.

A few comments:
- On the visibility of tasks, can we get together and collaborate. Primarily I want to look at near vs. long term (since you mentioned a different look).
- I suppose drill down is then related: is that opening the job and seeing the list of tasks in the current status or being able to filter the view to the tasks of a job?
- On the nesting of `jobs` (last point), I'm not sure that's a need - nor is necessarily using YAML/JSON. I guess I see it as `jobs` on one level and tasks at another (just as you illustrated). Thus, similar as you've added the dependency to the job level, we'd also want `tasks` to have dependencies within that job.

I'd add two things to the above:
- #35 would be really useful at task level as well.
- I would like to `tag` jobs/tasks. A `tag` at a job level would be visible to other jobs. A `tag` at a task level would only be visible to `tasks` within a particular job.
- Related to the `tag` above, we then want the ability to use a `tag` as a dependency (for example, all `tasks` within a particular job with `tag=Y` need to finish before a particular task executes)

**Comment by zeagler** [comment]
Most of the things noted above is included in the job feature of v0.2.1. 
https://github.com/zeagler/tangerine/releases/tag/v0.2.1

The only two things not addressed are:
* Task history - is not changed, every task still creates a new run. I think a separate issue is needed with a different approach for displaying history.
* #35, response to task failure, is not in the latest update.

**Comment by ManuelCalavera** [comment]
Awesome! It may be good to get a quick walk-through of all the latest features! Thanks
